File: uvaonlinejudge/src/volume_cxviii/P11800_DetermineTheShape.java
/*
 * MyPCS - My Programming Contests Solutions
 * 
 * Copyright (C) 2007 - 2011  Fernando Cardoso
 * email: fernandohbc@gmail.com / twitter: fernando_hbc
 * 
 * The source code of this program is free software; you can do whatever you
 * want with it since you always keep the name and contact of the original
 * author. Also, if you redistribute a modified version, make sure you explicit
 * the portions of the code that have been modified.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. In other words, use it
 * at your own risk.
 * 
 * If you find this a little useful, you can buy the author some beer.
 * If you find this really useful, you can buy a kid a book, and help making
 * the world more literate.
 */
package volume_cxviii;

import java.awt.geom.Line2D;
import java.awt.geom.Point2D;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

/**
 * UVa Online Judge Problem #11800 - Determine the Shape
 * 
 * @author Fernando Cardoso (fernandohbc@gmail.com)
 */
public class P11800_DetermineTheShape {
    private static final double EPSILON = 1e-15;
    private static long parsingPoints;
    private static long mountingOutput;
    private static long classifying;
    private static long printing;
    private static long readingPoints;

    public static void main(String[] args) {
        StringBuilder sb = new StringBuilder();
        Scanner scn = new Scanner(System.in);
        long ini = System.currentTimeMillis();
        long end;
        int q = scn.nextInt();
        for (int i = 1; i <= q; i++) {
            List<Point2D> points = new ArrayList<Point2D>();
            List<Line2D> segments = new ArrayList<Line2D>();
            for (int p = 0; p < 4; p++) {
                Point2D point = new Point2D.Float(scn.nextInt(), scn.nextInt());

                end = System.currentTimeMillis();
                readingPoints += (end - ini);
                ini = end;

                point_loop: for (Point2D inTheList : points) {
                    Line2D segment = new Line2D.Float(point, inTheList);
                    for (int s = 0; s < segments.size();) {
                        Line2D currentSegment = segments.get(s);
                        if (areCrossing(segment, currentSegment)) {
                            segments.remove(s);
                            continue point_loop;
                        } else {
                            s++;
                        }
                    }
                    segments.add(segment);
                }
                points.add(point);
            }

            end = System.currentTimeMillis();
            parsingPoints += (end - ini);
            ini = end;

            // print(segments);
            sb.append("Case ");
            sb.append(i);
            sb.append(": ");

            end = System.currentTimeMillis();
            mountingOutput += (end - ini);
            ini = end;

            String clazz = classify(segments);

            end = System.currentTimeMillis();
            classifying += (end - ini);
            ini = end;

            sb.append(clazz);
            sb.append('\n');

            end = System.currentTimeMillis();
            mountingOutput += (end - ini);
            ini = end;
        }
        System.out.println(sb.toString());
        end = System.currentTimeMillis();
        printing += (end - ini);

        System.out.println("Reading: " + readingPoints);
        System.out.println("Parsing: " + parsingPoints);
        System.out.println("Classifying: " + classifying);
        System.out.println("Mounting: " + mountingOutput);
        System.out.println("Printing: " + printing);

    }

    private static String classify(List<Line2D> segments) {
        segments = arrangeSegments(segments);
        // print(segments);
        double sizes[] = getSizes(segments);
        // System.out.println(Arrays.toString(sizes));
        double angles[] = getAngles(segments, sizes);
        // System.out.println(Arrays.toString(angles));

        if (allSame(sizes)) {
            if (allSame(angles)) {
                return "Square";
            } else {
                return "Rhombus";
            }
        }

        if (allSame(angles)) {
            return "Rectangle";
        }

        if (sizes[0] == sizes[2] && sizes[1] == sizes[3]) {
            return "Parallelogram";
        }

        if (Math.abs(Math.sin(getAngle(segments.get(0), segments.get(2),
                sizes[0], sizes[2]))) < EPSILON
                || Math.abs(Math.sin(getAngle(segments.get(1), segments.get(3),
                        sizes[1], sizes[3]))) < EPSILON) {
            return "Trapezium";
        }

        return "Ordinary Quadrilateral";

    }

    private static double getAngle(Line2D l1, Line2D l2, double d1, double d2) {
        double pi = (l1.getX2() - l1.getX1()) * (l2.getX2() - l2.getX1())
                + (l1.getY2() - l1.getY1()) * (l2.getY2() - l2.getY1());
        double angle = Math.acos(pi / (d1 * d2));
        return angle;
    }

    private static boolean allSame(double[] d) {
        return d[0] == d[1] && d[0] == d[2] && d[0] == d[3];
    }

    private static double[] getAngles(List<Line2D> segments, double[] sizes) {
        double[] result = new double[4];
        for (int i = 0; i < 4; i++) {
            result[i] = getAngle(segments.get(i), segments.get((i + 1) % 4),
                    sizes[i], sizes[(i + 1) % 4]);
        }
        return result;
    }

    private static double[] getSizes(List<Line2D> segments) {
        double[] result = new double[4];
        for (int i = 0; i < 4; i++) {
            result[i] = segments.get(i).getP1()
                    .distance(segments.get(i).getP2());
        }
        return result;
    }

    private static List<Line2D> arrangeSegments(List<Line2D> segments) {
        ArrayList<Line2D> arranged = new ArrayList<Line2D>();
        Line2D segment = segments.remove(0);
        arranged.add(segment);
        for (int i = 0; i < 3; i++) {
            Point2D begin = segment.getP1();
            Point2D end = segment.getP2();
            for (int s = 0; s < segments.size(); s++) {
                Line2D next = segments.get(s);
                if (next.getP1().equals(end) || next.getP2().equals(end)
                        || next.getP1().equals(begin)
                        || next.getP2().equals(begin)) {
                    segment = next;
                    arranged.add(next);
                    segments.remove(s);
                    break;
                }
            }
        }
        return arranged;
    }

    private static boolean areCrossing(Line2D s1, Line2D s2) {
        return s1.intersectsLine(s2) && !s1.getP1().equals(s2.getP1())
                && !s1.getP1().equals(s2.getP2())
                && !s1.getP2().equals(s2.getP1())
                && !s1.getP2().equals(s2.getP2());
    }
}

File: uvaonlinejudge/src/volume_cxviii/P11800_DetermineTheShape.cpp
/*
 * MyPCS - My Programming Contests Solutions
 * 
 * Copyright (C) 2007 - 2011  Fernando Cardoso
 * email: fernandohbc@gmail.com / twitter: fernando_hbc
 * 
 * The source code of this program is free software; you can do whatever you
 * want with it since you always keep the name and contact of the original
 * author. Also, if you redistribute a modified version, make sure you explicit
 * the portions of the code that have been modified.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. In other words, use it
 * at your own risk.
 * 
 * If you find this a little useful, you can buy the author some beer.
 * If you find this really useful, you can buy a kid a book, and help making
 * the world more literate.
 *
 *
 *
 * UVa Online Judge Problem #11800 - Determine the Shape
 * 
 * Author: Fernando Cardoso (fernandohbc@gmail.com)
 */

#include <iostream>
#include <vector>
#include <math.h>

#define EPSILON  1e-15

using namespace std;

struct Point {
  int x; int y;
};

struct Segment {
  Point p1; Point p2;
};

double getAngle(Segment s1, Segment s2, double norm1, double norm2) {
  double pi = (s1.p1.x - s1.p2.x)*(s2.p1.x-s2.p2.x) + (s1.p1.y-s1.p2.y)*(s2.p1.y-s2.p2.y);
  double angle = acos(pi / (norm1 * norm2));
  return angle;
}

bool allSame(double d[]) {
  return d[0] == d[1] && d[0] == d[2] && d[0] == d[3];
}

void getAngles(Segment segments[], double sizes[], double angles[]) {
  for (int i = 0; i < 4; i++) {
    angles[i] = getAngle(segments[i], segments[(i + 1) % 4], sizes[i], sizes[(i + 1) % 4]);
  }
}

void getSizes(Segment segments[], double sizes[]) {
  for (int i = 0; i < 4; i++) {
    sizes[i] = (segments[i].p1.x-segments[i].p2.x)*(segments[i].p1.x-segments[i].p2.x);
    sizes[i] += (segments[i].p1.y-segments[i].p2.y)*(segments[i].p1.y-segments[i].p2.y);
    sizes[i] = sqrt(sizes[i]);
  }
}

bool equals(Point p1, Point p2) {
  return p1.x == p2.x && p1.y == p2.y;
}

bool isAdjacent(Segment s1, Segment s2) {
  return equals(s1.p1, s2.p1) || equals(s1.p1, s2.p2) || equals(s1.p2, s2.p1) || equals(s1.p2, s2.p2);
}

void arrangeSegments(Segment segments[]) {
  for ( int i = 0; i <=2; i++ ) {
    int k = i + 1;
    while (!isAdjacent(segments[i], segments[k])) {
      k++;
    }
    Segment temp = segments[i+1];
    segments[i+1] = segments[k];
    segments[k] = temp;
  }
}

Point minusPoints(Point p1, Point p2) {
  Point result;
  result.x = p1.x-p2.x;
  result.y = p1.y-p2.y;
  return result;
}

double dot(Point p1, Point p2) {
  return p1.x*p2.x+p1.y*p2.y;
}

bool areCrossing(Segment ab, Segment cd) {
  Point e = minusPoints(ab.p2, ab.p1);
  Point f = minusPoints(cd.p2, cd.p1);

  Point p;
  p.x = -e.y;
  p.y = e.x;

  double  h = dot(minusPoints(ab.p1,cd.p1), p) / dot(f, p);

  return h > 0 && h < 1;
}

double abs(double d) {
  return d>=0?d:-d;
}

void print(Point p) {
  cout << "(" << p.x << ", " << p.y << ")";
}

void print(Segment s) {
  cout << "{";
  print(s.p1);
  cout << ", ";
  print(s.p2);
  cout << "}";
}

void print(Segment s[]) {
  cout << "[";
  for (int i = 0; i < 4; i++ ) {
    print(s[i]);
    if ( i < 3 ) {
      cout << "; ";
    }
  }
  cout << "]" << endl;

}

void print(double d[]) {
  cout << "[";
  for (int i = 0; i < 4; i++ ) {
    cout << d[i];
    if ( i < 3 ) {
      cout << ", ";
    }
  }
  cout << "]" << endl;
}

bool isVertical(Segment s) {
  return s.p1.x == s.p2.x;
}

double slope(Segment s) {
  // print(s);
  double result = 1.0*(s.p2.y-s.p1.y)/(s.p2.x-s.p1.x);
//  cout << " slope is " << result << endl;
  return result; 
}

bool parallel(Segment s1, Segment s2) {
  bool result;
  if ( isVertical(s1) ) {
    result = isVertical(s2);
  } else if ( isVertical(s2) ) {
    result = false;
  } else {
    result = slope(s1) == slope(s2);
  }
  
  //if ( result ) {
  //  print(s1);
  //  cout << " and ";
  //  print(s2);
  //  cout << " are parallel.";
  //}

  return result;
}

void classify(Segment segments[]) {
//  print(segments);
  arrangeSegments(segments);
//  print(segments);
  
  double sizes[4];
  getSizes(segments, sizes);
//  print(sizes);

  double angles[4];
  getAngles(segments, sizes, angles);
  // print(angles);

  if (allSame(sizes)) {
    if (allSame(angles)) {
      cout << "Square";
      return;
    } else {
      cout << "Rhombus";
      return;
    }
  }

 if (allSame(angles)) {
   cout << "Rectangle"; return;
 }

 if (sizes[0] == sizes[2] && sizes[1] == sizes[3]) {
   cout << "Parallelogram"; return;
 }

  if ( parallel( segments[0], segments[2]) || parallel(segments[1], segments[3])) {
    cout << "Trapezium"; return;
  }
  cout << "Ordinary Quadrilateral";
}



int main() {
  int q;
  cin >> q;
  for (int i = 1; i <= q; i++) {
    Point points[4];
    vector<Segment> segments;

    for (int p = 0; p < 4; p++) {
      Point point;

      cin >> point.x;
      cin >> point.y;
      for (int pi = 0; pi < p; pi++ ) {
        Segment segment;

        segment.p1 = points[pi];
        segment.p2 = point;

        bool crossing = false;
        for (int s = 0; s < segments.size() && !crossing;) {
          Segment currentSegment = segments[s];
          if (areCrossing(segment, currentSegment)) {
            segments.erase(segments.begin()+s);
            crossing = true;
          } else {
            s++;
          }
        }
        if ( !crossing ) {
          segments.push_back(segment);
        }
      }
      points[p] = point;
    }

    Segment segmentArray[4];
    for ( int s = 0; s < 4; s++ ) {
      segmentArray[s] = segments[s];
    }
    
    cout << "Case " << i << ": ";
    classify(segmentArray);
    cout << endl;
  }
}
